namer/core/src/main/scala/io/buoyant/namer/NameTreeTransformer.scala

```diff
@@ -22,6 +22,27 @@ trait NameTreeTransformer {
     override def bind(dtab: Dtab, path: Path): Activity[NameTree[Bound]] =
       underlying.bind(dtab, path).flatMap(transform)
   }
+
+  def wrap(underlying: Namer): Namer = new Namer {
+    private[this] def isBound(tree: NameTree[Name]): Boolean = {
+      tree match {
+        case NameTree.Neg | NameTree.Empty | NameTree.Fail => true
+        case NameTree.Alt(trees@_*) => trees.forall(isBound)
+        case NameTree.Union(trees@_*) => trees.map(_.tree).forall(isBound)
+        case NameTree.Leaf(_: Name.Bound) => true
+        case NameTree.Leaf(_) => false
+      }
+    }
```

Comment:
it seems like this could be done as `checkBound(tree: NameTree[Name]): Option[NameTree[Name.Bound]]` to avoid the casting.  But I don't really see this as a severe issue...